numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -1750,12 +1802,133 @@ _array_fromobject(PyObject *NPY_UNUSED(ignored), PyObject *args, PyObject *kws)
      * steals a reference to ret
      */
     return _prepend_ones(ret, nd, ndmin, order);
+}
 
-clean_type:
-    Py_XDECREF(type);
-    return NULL;
+#undef LONG_CONVERTER
+#undef FAST_PARSE
+#undef STRIDING_OK
+
+/*
+ * The following functions define the array coercion functions from python.
+ * They do manual parsing since they are ubiquitous and the micro-optimization
+ * thus is worth the trouble.
+ */
+static PyObject *
+array_array(PyObject *NPY_UNUSED(ignored), PyObject *args, PyObject *kwargs)
```

Comment:
I'd be tempted to put these in an `as_array.c` file just to keep them separated like they were when they were in python.

Reply:
Was thinking about it as well. But maybe should discuss cython briefly first also then. @mattip I do not recall any discussion, although I am not sure I like it. The `as...` functions are usually used without many arguments. If it gets more complex, there seems little reason not to use `np.array()` directly.